server/queue/queue-api/src/test/java/org/apache/james/queue/api/MailQueueContract.java

```diff
@@ -0,0 +1,286 @@
+            .isEqualTo(state);
+    }
+
+    @Test
+    default void queueShouldPreserveRemoteAddress() throws Exception {
+        String remoteAddress = "remote";
+        getMailQueue().enQueue(defaultMail()
+            .remoteAddr(remoteAddress)
+            .build());
+
+        MailQueue.MailQueueItem mailQueueItem = getMailQueue().deQueue();
+        assertThat(mailQueueItem.getMail().getRemoteAddr())
+            .isEqualTo(remoteAddress);
+    }
+
+    @Test
+    default void queueShouldPreserveRemoteHost() throws Exception {
+        String remoteHost = "remote";
+        getMailQueue().enQueue(defaultMail()
+            .remoteHost(remoteHost)
+            .build());
+
+        MailQueue.MailQueueItem mailQueueItem = getMailQueue().deQueue();
+        assertThat(mailQueueItem.getMail().getRemoteHost())
+            .isEqualTo(remoteHost);
+    }
+
+    @Test
+    default void queueShouldPreserveLastUpdated() throws Exception {
+        Date lastUpdated = new Date();
+        getMailQueue().enQueue(defaultMail()
+            .lastUpdated(lastUpdated)
+            .build());
+
+        MailQueue.MailQueueItem mailQueueItem = getMailQueue().deQueue();
+        assertThat(mailQueueItem.getMail().getLastUpdated())
+            .isEqualTo(lastUpdated);
+    }
+
+    @Test
+    default void queueShouldPreserveName() throws Exception {
+        getMailQueue().enQueue(defaultMail()
+            .name("name")
+            .build());
+
+        MailQueue.MailQueueItem mailQueueItem = getMailQueue().deQueue();
+        assertThat(mailQueueItem.getMail().getName())
+            .isEqualTo("name");
+    }
+
+    @Test
+    default void queueShouldPreservePerRecipientHeaders() throws Exception {
+        PerRecipientHeaders.Header header = PerRecipientHeaders.Header.builder()
+            .name("any")
+            .value("any")
+            .build();
+        getMailQueue().enQueue(defaultMail()
+            .addHeaderForRecipient(header, RECIPIENT1)
+            .build());
+
+        MailQueue.MailQueueItem mailQueueItem = getMailQueue().deQueue();
+        assertThat(mailQueueItem.getMail().getPerRecipientSpecificHeaders()
+            .getHeadersForRecipient(RECIPIENT1))
+            .containsOnly(header);
+    }
+
+    @Test
+    default void dequeueShouldBeFifo() throws Exception {
+        getMailQueue().enQueue(defaultMail()
+            .name("name1")
+            .build());
+        getMailQueue().enQueue(defaultMail()
+            .name("name2")
+            .build());
+
+        MailQueue.MailQueueItem mailQueueItem1 = getMailQueue().deQueue();
+        mailQueueItem1.done(true);
+        MailQueue.MailQueueItem mailQueueItem2 = getMailQueue().deQueue();
+        mailQueueItem2.done(true);
+        assertThat(mailQueueItem1.getMail().getName()).isEqualTo("name1");
+        assertThat(mailQueueItem2.getMail().getName()).isEqualTo("name2");
+    }
+
+    @Test
+    default void dequeueCouldBeInterleaving() throws Exception {
```

Comment:
what is supposed to be "invertleaving" ?

Reply:
Maybe "dequeueCanBeChainedBeforeAck"

Reply:
great